source/htslib-1.2.1/htsfile.1
.TH htsfile 1 "3 February 2015" "htslib-1.2.1" "Bioinformatics tools"
.SH NAME
htsfile \- identify high-throughput sequencing data files
.\"
.\" Copyright (C) 2015 Genome Research Ltd.
.\"
.\" Author: John Marshall <jm18@sanger.ac.uk>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software is furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
.\" THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\" FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\" DEALINGS IN THE SOFTWARE.
.\"
.SH SYNOPSIS
.B htsfile
.RB [ -chH ]
.IR FILE ...
.SH DESCRIPTION
The \fBhtsfile\fR utility attempts to identify what kind of high-throughput
sequencing data files the specified files are, and provides minimal viewing
capabilities for some kinds of data file.
.P
It can identify sequencing data files such as SAM, BAM, and CRAM;
variant calling data files such as VCF and BCF;
index files used to index these data files;
and compressed versions of many of them.
.P
For each \fIFILE\fR given, \fBhtsfile\fP prints a description of the file
format determined, using similar keyword conventions to \fBfile\fP(1):
"text" indicates a textual file that can probably be viewed on a terminal;
"data" indicates binary data;
"sequence", "variant calling", and "index" indicate different categories of
data file.
When it can be identified, the name of the particular file format (such as
"BAM" or "VCF") is printed at the start of the description.
.P
When used to view file contents as text, \fBhtsfile\fP can optionally show
only headers or only data records, but has no other filtering capabilities.
Use \fBsamtools\fR or \fBbcftools\fR if you need more extensive viewing or
filtering capabilities.
.P
The following options are accepted:
.TP 4n
.BR -c ", " --view
Instead of identifying the specified files, display a textual representation
of their contents on standard output.
.TP
.BR -h ", " --header-only
Display data file headers only.
Implies \fB--view\fR.
.TP
.BR -H ", " --no-header
When viewing files, display data records only.
.PP
.SH SEE ALSO
.IR bcftools (1),
.IR file (1),
.IR samtools (1)


CHANGES
Implemented --chimSegmentReadGapMax parameter which defines the maximum gap in the read sequence between chimeric segments. By default it is set to 0 to replicate the behavior of the previous STAR versions.
Implemented --chimFilter banGenomicN | None options to prohibit or allow the N characters in the vicinity of the chimeric junctions. By default, they are prohibited - the same behavior as in the previous versions.

STAR 2.4.2a 2015/06/19
Implemented --quantMode GeneCounts option for counting number of reads per gene, similar to htseq-count.
STARlong: fixed --outFilterIntronMotifs and --outSAMstrandField options.
Yet another fix for --sjdbOverhang logic.
Error message when shared memory and on the fly junction insertion are used together.
Fixed a bug causing unnecessary 1 base soft-clipping in rare cases with sparse suffix array.
Fixed a bug that caused problems with junction motifs in rare cases. Very few alignments affected, <1 per million.

STAR 2.4.1d 2015/05/19
Fixed problems with --sjdbOverhang default and user-defined values.
Fixed problems with occasional non-adjacent output of multiple alignments into the unsorted BAM file and transcriptome BAM file.
Fixed a bug causing seg-faults when shared memory options in --genomeLoad are used with --outStd SAM.
Fixed a bug causing seg-faults for small values of --limitIObufferSize.
Added STAR long pre-compiled executables.
Fixed very minor issues with filtering into SJ.out.tab .
Fixed some bugs in STARlong mapping algorithm.
Fixed --outFilter BySJout filtering for STARlong.
Fixed XS attrbutes in STARlong.
Added --runDirPerm option for permissions of run-time directories.


STAR 2.4.1c 2015/04/24
Added latest version of STAR-Fusion as a separate directory.
Fixed some compilation problems introduced in 2.4.1b.
Added Mac executable.


STAR 2.4.1b 2015/04/23
Fixed a bug introduced in 2.4.1a causing serious problems for genomes generated without annotations.
      If you generated a genome without annotations with 2.4.1a please re-generate it.
Fixed a bug causing seg-faults when generating genomes with a large (>500k) number of junctions.
Fixed a bug causing seg-faults with --chimOutType WithinBAM for single-end reads.
Fixed a bug with required --sjdbOverhang at the mapping step.

STAR 2.4.1a 2015/04/17
The annotations can now be included on the fly at the mapping step, without including them at the genome generation step.
New option to activate on the fly "per sample" 2-pass method: "--twopassMode Basic".
2-pass mode can now be used with annotations, which can be included either at the run-time, or at the genome generation step.
Included link (submodule) to Brian Haas' STAR-Fusion code for detecting fusion transcript from STAR chimeric output:  https://github.com/STAR-Fusion/STAR-Fusion
Included Gery Vessere's shared memory implementation for POSIX and SysV. To compile STAR with POSIX shared memory, use `make POSIXSHARED`
New option "--chimOutType WithinBAM" to include chimeric alignments together with normal alignments in the main (sorted or unsorted) BAM file(s).
New option "--quantTranscriptomeBan Singleend" allows insertions, deletions ans soft-clips in the transcriptomic alignments, which are allowed by some expression quantification software (e.g. eXpress). 
New option "--alignEndsTypeExtension Extend5pOfRead1" to enforce full extension of the 5p of the read1, while all other ends undergo local alignment and may be soft-clipped.

2.4.0k 03/30/2015
Implemented new BAM sorting algorithm that reduces significantly the required RAM.

2.4.0j 02/04/2015
Fixed a problem with scoring alignments for STARlong. STARlong alignments are slightly modified.
Fixed a bug introduced in 2.4.0i that dropped a large number of aligmnents for --quantMode TranscriptomeSAM.
	Transcriptome alignments are now the same as in version 2.4.0h.
Fixed a problem with lower case read sequences for --outSAMtype BAM options.
Fixed a bug preventing parameter value to be "-".
Fixed --genomeLoad LoadAndRemove option.

2.4.0i 01/14/2015
Fixed a bug with the _STARtmp temporary directory name for the 2-pass runs.
Fixed a bug causing seg-faults for genome generation.
Fixed a bug causing seg-faults for --quantMode TranscriptomeSAM

2.4.0h 12/09/2014
Fixed the problem causing Ubuntu error: "sh: 1: Syntax error: Bad fd number".
Added --quantTranscriptomeBAMcompression option.
Add newline at the end of STAR_VERSION string (contributed by Daniel Nicorici).
Fixed a bug with parsing the last line of paired FASTA files (contributed by Alex Rolfe).

2.4.0g 11/26/2014
Fixed a bug with output score (AS attribute) of some chimeric alignments.
Added --alignSoftClipAtReferenceEnds No option which prevents soft clipping of alignments at the reference (chromosome) ends, for compatibility with Cufflinks/Cuffmerge.
Fixed wrong output of chimeric junctions boundaries in the Chimeric.out.junction file.
Added --outSAMflagOR, --outSAMflagAND options to set specific bits of the SAM FLAG.
--sjdbFileChrStartEnd can now accept multiple files which will be concatenated.
Fixed the header of the Log.progress.out .
Fixed a bug that prevented output of transcriptomic alignments (--quantMode TranscriptomeSAM)  with 1 base junction overhangs.
Added --sysShell option to specify path to bash, in cases where bash is not the default shell.
--outBAMcompression default changed to 1, which apparently does not change deflation ratio, but is much faster.
Added --outBAMsortingThreadN option to specify number of threads for BAM sorting. By default (0) it's equal to min(6,runThreadN).

2.4.0f1 10/30/2014
Added read group (RG) BAM attributes to transcriptome BAM (contributed by https://github.com/godotgildor).
Fixed a bug with double ID field in the RG header line (contributed by https://github.com/godotgildor).
Fixed a bug in the 2-pass method (--twopass1readsN).
Fixed a problem with RAM allocation for BAM sorting.

2.4.0e 10/24/2014
Added manual in PDF.
New sub-directories: source, bin, doc.

source/htslib-1.2.1/htsfile.c
/*  htsfile.c -- file identifier and minimal viewer.

    Copyright (C) 2014-2015 Genome Research Ltd.

    Author: John Marshall <jm18@sanger.ac.uk>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.  */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#ifdef _WIN32
#include "common\getopt\getopt.h"
#else
#include <getopt.h>
#endif
#include <unistd.h>
#include "common\cross_platform.h"
#include "htslib/hfile.h"
#include "htslib/hts.h"
#include "htslib/sam.h"
#include "htslib/vcf.h"

enum { identify, view_headers, view_all } mode = identify;
int show_headers = 1;

static htsFile *dup_stdout(const char *mode)
{
    int fd = dup(STDOUT_FILENO);
    hFILE *hfp = (fd >= 0)? hdopen(fd, mode) : NULL;
    return hfp? hts_hopen(hfp, "-", mode) : NULL;
}

static int view_sam(hFILE *hfp, const char *filename)
{
    samFile *in = hts_hopen(hfp, filename, "r");
    if (in == NULL) return 0;
    samFile *out = dup_stdout("w");
    bam_hdr_t *hdr = sam_hdr_read(in);

    if (show_headers) sam_hdr_write(out, hdr);
    if (mode == view_all) {
        bam1_t *b = bam_init1();
        while (sam_read1(in, hdr, b) >= 0)
            sam_write1(out, hdr, b);
        bam_destroy1(b);
    }

    bam_hdr_destroy(hdr);
    hts_close(out);
    hts_close(in);
    return 1;
}

static int view_vcf(hFILE *hfp, const char *filename)
{
    vcfFile *in = hts_hopen(hfp, filename, "r");
    if (in == NULL) return 0;
    vcfFile *out = dup_stdout("w");
    bcf_hdr_t *hdr = bcf_hdr_read(in);

    if (show_headers) bcf_hdr_write(out, hdr);
    if (mode == view_all) {
        bcf1_t *rec = bcf_init();
        while (bcf_read(in, hdr, rec) >= 0)
            bcf_write(out, hdr, rec);
        bcf_destroy(rec);
    }

    bcf_hdr_destroy(hdr);
    hts_close(out);
    hts_close(in);
    return 1;
}

static void usage(FILE *fp, int status)
{
    fprintf(fp,
"Usage: htsfile [-chH] FILE...\n"
"Options:\n"
"  -c, --view         Write textual form of FILEs to standard output\n"
"  -h, --header-only  Display only headers in view mode, not records\n"
"  -H, --no-header    Suppress header display in view mode\n");
    exit(status);
}

int main(int argc, char **argv)
{
    static const struct option options[] = {
        { "header-only", no_argument, NULL, 'h' },
        { "no-header", no_argument, NULL, 'H' },
        { "view", no_argument, NULL, 'c' },
        { "help", no_argument, NULL, '?' },
        { "version", no_argument, NULL, 1 },
        { NULL, 0, NULL, 0 }
    };

    int status = EXIT_SUCCESS;
    int c, i;
    while ((c = getopt_long(argc, argv, "chH?", options, NULL)) >= 0)
        switch (c) {
        case 'c': mode = view_all; break;
        case 'h': mode = view_headers; show_headers = 1; break;
        case 'H': show_headers = 0; break;
        case 1:
            printf(
"htsfile (htslib) %s\n"
"Copyright (C) 2015 Genome Research Ltd.\n",
                   hts_version());
            exit(EXIT_SUCCESS);
            break;
        case '?': usage(stdout, EXIT_SUCCESS); break;
        default:  usage(stderr, EXIT_FAILURE); break;
        }

    if (optind == argc) usage(stderr, EXIT_FAILURE);

    for (i = optind; i < argc; i++) {
        htsFormat fmt;
        hFILE *fp = hopen(argv[i], "r");
        if (fp == NULL) {
            fprintf(stderr, "htsfile: can't open \"%s\": %s\n", argv[i], strerror(errno));
            status = EXIT_FAILURE;
            continue;
        }

        if (hts_detect_format(fp, &fmt) < 0) {
            fprintf(stderr, "htsfile: detecting \"%s\" format failed: %s\n", argv[i], strerror(errno));
            hclose_abruptly(fp);
            status = EXIT_FAILURE;
            continue;
        }

        if (mode == identify) {
            char *description = hts_format_description(&fmt);
            printf("%s:\t%s\n", argv[i], description);
            free(description);
        }
        else
            switch (fmt.category) {
            case sequence_data: if (view_sam(fp, argv[i])) fp = NULL; break;
            case variant_data:  if (view_vcf(fp, argv[i])) fp = NULL; break;
            default:
                fprintf(stderr, "htsfile: can't view %s: unknown format\n", argv[i]);
                status = EXIT_FAILURE;
                break;
            }

        if (fp && hclose(fp) < 0) {
            fprintf(stderr, "htsfile: closing %s failed\n", argv[i]);
            status = EXIT_FAILURE;
        }
    }

    return status;
}
